### PR TITLE
Permissions

### PR DESCRIPTION
**Pull request by luisdemarchi**
Created the full stream for permissions.

### CODE REVIEW

**Comment by kosiraljaz** [comment]
Is it possible to also enable permissions for sending messages. So for example only users that belong to a specific group will receive the message? Currently it's only possible to deny users at subscribe level. Do you have any plans for this?

Thanks for your work.

**Comment by luisdemarchi** [comment]
@kosiraljaz The code that I have ready for my use, contemplates the selection of which group should go the answer of each class (I believe that here it will be possible to make the rule that you want). But since this is the first project I'm working on, I'll wait for the permissions to be reviewed and then add that other part of the code.

I'm sorry for my English, I'm in Brazil.

File: channels_api/bindings.py [hunk 1]
```diff
@@ -78,8 +78,13 @@ def _group_name(self, action, id=None):
...
 
-    def has_permission(self, action, pk, data):
-        return True
+    def has_permission(self, user, action, pk):
+        permissions = True
+
+        for class_permission in self.permission_classes:
+            permission = class_permission().has_permission(user, action, pk)
+            permissions = permission and permissions
+        return permissions
```
Comment (theY4Kman) [inline channels_api/bindings.py]: This loop could short-circuit, so it returns False at first permission denial:
```python
def has_permission(self, user, action, pk):
    for class_permission in self.permission_classes:
        if not class_permission().has_permission(user, action, pk):
            return False
    else:
        return True
```

Comment (luisdemarchi) [inline channels_api/bindings.py]: You're right. Thank you. Sorry for the ignorance, but do I need to do something or was that included automatically?

Comment (theY4Kman) [inline channels_api/bindings.py]: If you want to make the change, you'll have to make another commit and push it as well.

File: tox.ini [hunk 2]
```diff
@@ -1,17 +1,14 @@
 [tox]
 envlist =
-  {py27}-django-{18,19,110}
-  {py34}-django-{18,19,110}
-  {py35}-django-{18,19,110}
+  {py34}-django-{110}
```
Comment (linuxlewis) [inline tox.ini]: These tox.ini changes are incorrect. They should match master

File: tests/models.py [hunk 3]
```diff
@@ -1,7 +1,6 @@
 from django.db import models
```
Comment (linuxlewis) [inline tests/models.py]: Remove these changes please

File: channels_api/routing.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from django.core.serializers.json import json
```
Comment (linuxlewis) [inline channels_api/routing.py]: I don't think this module is actually needed. I would expect the test code to just use the standard channels.auth decorators like `channel_session_user_from_http`

**Comment by Mehotkhan** [comment]
hi
please merge this pull request

**Comment by luisdemarchi** [comment]
Following is the actual structure:

```
# Presence REDIS DB (no-SQL)
db = walrus.Database(host='localhost', port=6379, db=0)


class Presence(walrus.Model):
    database = db
    channel_name = walrus.TextField(primary_key=True)
    user = walrus.TextField(fts=True, index=True)
    last_seen = walrus.DateTimeField(default=datetime.datetime.now)
    is_accessible = walrus.BooleanField(index=True, default=False)
```

**Comment by linuxlewis** [comment]
Thanks for your submission closing this in favor of #18
